Appenderators, DataSource metadata, KafkaIndexTask

gianm opened the pull request:
See also #1642. FYI we are running this in our cluster as of 2016-03-02, things seem to work so far…

Set of three related things (each one a separate commit):

**kafka-indexing-service core extension**
Includes KafkaIndexTask, which reads a specific offset range from specific partitions, and can use dataSource metadata transactions to guarantee exactly-once ingestion.

Each task has a finite lifecycle, so it is expected that a process will be supervising existing tasks and creating new ones when needed. @dclim is working on this process in https://github.com/dclim/druid/tree/kafka-supervisor.

This extension requires the other two features (DataSource metadata and Appenderators).

**DataSource metadata**
Geared towards supporting transactional inserts of new segments. This involves an interface `DataSourceMetadata` that allows combining of partially specified metadata (useful for partitioned ingestion). It also involves changes to the `SegmentInsertAction` to allow it to take a `startMetadata` and `endMetadata` for compare-and-swap.

DataSource metadata is stored in a new "dataSource" table.

**Appenderators**
Like Plumbers, but different. Appenderators are a way of getting more control over the ingestion process than a Plumber allows. They are less ambitious than Plumbers, but more flexible. In particular, they offer facilities to deal with:

- Indexing and persisting data
- Checkpointing and restoring ingestion state
- Merging and pushing segments to deep storage
- Serving queries

But they do _not_ do any of these things:

- Decide which rows to index and which to drop
- Decide which segments to put your rows into
- Decide when to push and publish segments
- Monitor handoff

So you can think of Appenderators as a way of separating out the mechanical functionality of Plumbers from their decision-making processes. The idea is that existing Plumbers could be implemented using Appenderators, but you could also use them to implement workflows that the existing Plumbers can't support.

**Discussion?**
Some open questions (and reasons this is still marked Discuss),

- Should RealtimePlumber be adapted to use an Appenderator? Pros of doing this are that the `AppenderatorImpl` actually has a lot of functionality and code overlap with RealtimePlumber (the query-runner stuff is particularly annoying since it is mostly similar but also kinda different). I think the main stumbling block is that the RealtimePlumber and the AppenderatorImpl have different persist directory layouts (mostly because the RealtimePlumber never has to have more than one shard per interval, but the AppenderatorImpl might) and so the code will need to be able to migrate existing data.

- Should the other existing indexy things be adapted to use an Appenderator too? Like IndexTask and IndexGeneratorReducer.

BTW, I think if we do either of these things, they should be in future PRs rather than this one.

fjy commented:
I am so excited!

gianm commented:
[wip] because TODOs remain, but would really appreciate comments

himanshug commented on indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentInsertAction.java:
```diff
@@ -59,26 +72,41 @@ public SegmentInsertAction(
+      toolbox.getEmitter().emit(metricBuilder.build("segment/txn/success", 1));
+    } else {
+      toolbox.getEmitter().emit(metricBuilder.build("segment/txn/failure", 1));
+    }
```
i understand the usefulness of `segment/txn/failure` but not sure how `segment/txn/success` is going to be useful.

gianm replied:
maybe you want to confirm that txns are actually happening? I dunno? It seemed like if one existed then the other should too

himanshug commented on server/src/main/java/io/druid/indexing/overlord/IndexerMetadataStorageCoordinator.java:
```diff
@@ -93,9 +93,35 @@ SegmentIdentifier allocatePendingSegment(
+   * Attempts to insert a set of segments to the metadata storage. Returns the set of segments actually added (segments
+   * with identifiers already in the metadata storage will not be added).
+   * <p/>
+   * If newCommitMetadata is set, this insertion will be atomic with a compare-and-swap on dataSource commit metadata.
```
only start/end Metadata are there.

gianm replied:
good point, will fix the comment

himanshug commented on server/src/main/java/io/druid/indexing/overlord/SegmentPublishResult.java:
```diff
@@ -0,0 +1,95 @@
+ * Result of an operation that attempts to publish segments. Indicates the set of segments actually published
+ * and whether or not the transaction was a success.
+ *
+ * If "success" is false then the segments set will be empty.
```
can segments set be empty even when success is true? if not, why do we need the flat at all?

gianm replied:
yes, since "segments" only includes segments actually published in this transaction- it's possible for the same segments to have been published in some other transaction (like if there was a replica)

himanshug replied:
got it, thanks

gianm replied:
added this to the comment.

himanshug commented on server/src/main/java/io/druid/indexing/overlord/SegmentPublishResult.java:
```diff
@@ -0,0 +1,95 @@
+           ", success=" + success +
+           '}';
+  }
+}
```
newline

himanshug commented on server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java:
```diff
@@ -259,24 +266,88 @@ public void start()
+                if (announceHistoricalSegment(handle, segment)) {
+                  inserted.add(segment);
+                }
+              }
```
shouldn't we try and do it only after txn status has been successfully found/updated to be true?

gianm replied:
if the transaction fails, an empty set will be returned (`inserted` gets ignored)

himanshug replied:
yes, but why make db do all the work  (all calls to announceHistoricalSegment(handle, segment)) and rollback... we can do the check before to reduce the work that needs to be done at rollback.

gianm replied:
oh, I see what you're saying. Sure I can switch these

himanshug commented on server/src/main/java/io/druid/segment/realtime/appenderator/SegmentAllocator.java:
```diff
@@ -0,0 +1,36 @@
+   *
+   * @return the pending segment identifier, or null if it was impossible to allocate a new segment
+   */
+  SegmentIdentifier allocate(DateTime timestamp) throws IOException;
```
couldn't find any implementations for this.

gianm replied:
there's an anonymous implementation in KafkaIndexTask

himanshug replied:
thanks, failed to notice that... "quick scan".. duh

himanshug commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaDataSourceMetadata.java:
```diff
@@ -0,0 +1,114 @@
+
+      for (Map.Entry<Integer, Long> entry : that.getKafkaPartitions().getPartitionOffsetMap().entrySet()) {
+        newMap.put(entry.getKey(), entry.getValue());
+      }
```
should this check that value is greater and then only override?

gianm replied:
This method is used by `matches` to figure out if things match, so the overriding is good as it helps detect problems.

himanshug commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaDataSourceMetadata.java:
```diff
@@ -0,0 +1,114 @@
+      return new KafkaDataSourceMetadata(new KafkaPartitions(kafkaPartitions.getTopic(), newMap));
+    } else {
+      // Different topic, prefer "other".
+      return other;
```
this looks like and error case and should fail with "WTF"

gianm replied:
similarly to above, this method is used by `matches` to figure out if things match, so the overriding is good as it helps detect problems.

himanshug replied:
but this is not following the very explicit contract set by the interface, you could create another helper method that can be used by both plus and matches , then plus could maintain the contract and fail in this situation.
makes sense?

gianm replied:
I think it does follow the interface, the docs for `plus` say `Any conflicts should be resolved in favor of information from "other".` So that's what this line is doing, preferring "other". I don't really see a need for another method, as it's pretty easy for callers to use `matches` to determine if two metadatas are compatible, and then `plus` to add them if they are.

Am I missing something?

himanshug commented on server/src/main/java/io/druid/segment/realtime/appenderator/TransactionalSegmentPublisher.java:
```diff
@@ -0,0 +1,40 @@
+      Set<DataSegment> segments,
+      Object commitMetadata
+  ) throws IOException;
+}
```
no impls found

gianm replied:
@himanshug this has an anonymous impl in KafkaIndexTask

himanshug commented:
did a scan, looks like a step in right direction, waiting to see the impl for SegmentAllocator.
i'm not sure about the name kafka-indexing-service which gave me the notion that this was some kind of "indexing service", which rings me overlord bcoz of the renaming history

also, i believe, we should set the "scope" for this PR to at least result in a working (even if experimental) version of no-window period realtime ingestion with tranquility.

gianm commented:
@himanshug SegmentAllocator has an implementation in KafkaIndexTask.

Do you have a suggestion for a better name for kafka-indexing-service? I just called it that because it has Kafka stuff for the indexing service.

Hmm IMO this PR should be scoped to be the base for tranquility working and for kafka working. It actually doesn't fully achieve either one but it was getting big enough that I thought it made sense to cut it off at this point and do the rest in follow on PRs.

I think the follow on PRs would be:

1. One supporting Kafka that introduces a supervisor that actually spawns a series of KafkaIndexTasks
2. One supporting Tranquility that introduces a push-oriented task based on the FiniteAppenderatorDriver, which Tranquility can push to
3. A Tranquility PR (in the Tranquility repo) that coordinates that new task in (2)

himanshug commented:
we can rename kafka-indexing-service to just kafka-indexing or may be I'm overthinking?

regarding scope, I would move the kafka module in a separate PR and do (2) in this PR itself. That said, I will let you take the final call on that.

gianm commented:
@himanshug kafka-indexing sounds cool

Hmm, I would prefer to have this PR have the stuff it has for now, mostly because this stuff is "done" (ish) and the follow-on stuff is not done :). I did start working on (2) though, so depending on how long this PR is open for, we could potentially include that instead of the kafka stuff.

Out of curiosity why do you prefer to have the tranquility stuff here instead of the kafka stuff?

himanshug commented:
"Out of curiosity why do you prefer to have the tranquility stuff here instead of the kafka stuff?"
to get you focused on that and have stuff working for tranquility as early as possible.

also, I dint realize that (2) was going to be "tranquility specific" code, i thought that was core druid change and not a new module.

gianm commented:
Haha, ok :)

I actually had been doing the tranquility stuff first and then took a detour through the kafka stuff as I think it is a lot simpler, and wanted to get some kind of initial thing working. The tranquility stuff ended up getting kind of complicated. I plan to go back to that soon though. But in the meantime, some review on this stuff would be great.

himanshug commented on server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java:
```diff
@@ -0,0 +1,397 @@
+        final long sleepMillis = computeNextRetrySleep(++nTry);
+        log.warn(e, "Failed publishAll (try %d), retrying in %,dms.", nTry, sleepMillis);
+        Thread.sleep(sleepMillis);
+      }
```
various places where we have long running stuff, we need to check for thread interrupt status and finish early, so that processes stop properly.

gianm replied:
an interrupt will cause `publishAll` to throw an InterruptedException (due to the Thread.sleep), is that enough?

himanshug replied:
what if the Exception e caught here is an InterruptedException?

gianm replied:
good point, will fix.

himanshug commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java:
```diff
@@ -0,0 +1,479 @@
+          throw new InterruptedException();
+        }
+
+        final ConsumerRecords<byte[], byte[]> records = consumer.poll(POLL_TIMEOUT);
```
i guess poll(timeout) returns empty records on timeout, so the while loop is going to run indefinitely if there was no new data. we have to provide ways for early termination. i guess some of these things will need to be [re]visited as the "supervisor" takes shape.

gianm replied:
yeah, currently there is no way to early terminate. will add a TODO for that.

himanshug commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java:
```diff
@@ -0,0 +1,479 @@
+  {
+    // TODO
+  }
+
```
must be a restorable task and be able to stop gracefully. i know its todo :) , but just emphasizing.

gianm replied:
yep, agree :)

gianm replied:
pushed an implementation of graceful stop & restore.

himanshug commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIOConfig.java:
```diff
@@ -0,0 +1,113 @@
+    this.startPartitions = Preconditions.checkNotNull(startPartitions, "startPartitions");
+    this.endPartitions = Preconditions.checkNotNull(endPartitions, "endPartitions");
+    this.consumerProperties = Preconditions.checkNotNull(consumerProperties, "consumerProperties");
+    this.useTransaction = useTransaction;
```
this must be true by default.

gianm replied:
could make it `skipTransaction` then, and default to false.

gianm replied:
hmm, I think the double negatives make the code a bit weird, imo we could address this by making transactions the default at the supervisor level

himanshug replied:
can't you have ` @JsonProperty("useTransaction") Boolean useTransaction` in the argument and then you can do
```
this.useTransaction = useTransaction == null ? true : useTransaction;
```

gianm replied:
sure, ok.

fjy commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaDataSourceMetadata.java:
```diff
@@ -0,0 +1,114 @@
+  @Override
+  public DataSourceMetadata plus(DataSourceMetadata other)
+  {
+    final KafkaDataSourceMetadata that = (KafkaDataSourceMetadata) other;
```
can we have a nice error msg in case the cast fails?

gianm replied:
ok

fjy commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIOConfig.java:
```diff
@@ -0,0 +1,113 @@
+
+  @JsonCreator
+  public KafkaIOConfig(
+      @JsonProperty("sequenceName") String sequenceName,
```
docs needed for what these mean and how to usethem

gianm replied:
this is an internal impl detail and I do not think it needs to be documented. the supervisor is the user-facing thing.

fjy commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java:
```diff
@@ -0,0 +1,526 @@
+    this.ioConfig = Preconditions.checkNotNull(ioConfig, "ioConfig");
+  }
+
+  private static String makeTaskId(String dataSource, int randomBits)
```
can this be extracted out somewhere and reused for other tasks?

fjy commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java:
```diff
@@ -0,0 +1,526 @@
+  }
+
+  @JsonProperty
+  public KafkaIOConfig getIoConfig()
```
IOConfig? or IoConfig?

gianm replied:
changed

fjy commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java:
```diff
@@ -0,0 +1,526 @@
+    );
+
+    try (
+        final Appenderator appenderator0 = newAppenderator(fireDepartmentMetrics, toolbox);
```
firstAppenderator?

gianm replied:
they're the same appenderator, just some annoying try-with-resources stuff going on here.

fjy commented on indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentInsertAction.java:
```diff
@@ -40,17 +41,29 @@
-      @JsonProperty("segments") Set<DataSegment> segments
+      @JsonProperty("segments") Set<DataSegment> segments,
+      @JsonProperty("startMetadata") DataSourceMetadata startMetadata,
+      @JsonProperty("endMetadata") DataSourceMetadata endMetadata
```
are there jackson annotated getters for these variables?

dclim replied:
+1, serialization loses startMetadata/endMetadata and entries are not written into druid_dataSource table.

gianm replied:
fixed

fjy commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1192 @@
+  private final ServiceEmitter emitter;
+  private final ExecutorService queryExecutorService;
+  private final IndexIO indexIO;
+  private final IndexMerger indexMerger;
```
does this support v9 indexing?

gianm replied:
yes, it's up to the caller to pass in the desired impl of IndexMerger.

fjy commented on server/src/main/java/io/druid/segment/realtime/plumber/Sink.java:
```diff
@@ -126,6 +131,10 @@ public int add(InputRow row) throws IndexSizeExceededException
 
     synchronized (hydrantLock) {
+      if (!writable) {
+        return -1;
```
we should make this into a constant

gianm replied:
ok, will do.

fjy commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1192 @@
+      sinkRetVal = sink.add(row);
+    }
+    catch (IndexSizeExceededException e) {
+      // Try one more time after swapping, then throw the exception out if it happens again.
```
why once?

gianm replied:
if we swap the index out, and the new, empty index *still* can't hold the row, then the row must be really gigantor and really cannot be indexed.

fjy commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1192 @@
+      );
+
+      try {
+        segmentAnnouncer.announceSegment(retVal.getSegment());
```
it feels really weird that a get() method will announce the segment

gianm replied:
renamed.

fjy commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1192 @@
+    }
+  }
+
+  private Sink getSink(final SegmentIdentifier identifier)
```
i think this needs a differet name

gianm replied:
any suggestions?

gianm replied:
renamed to `getOrCreateSink`

fjy commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1192 @@
+              @Override
+              public Iterable<TimelineObjectHolder<String, Sink>> apply(final Interval interval)
+              {
+                return sinkTimeline.lookup(interval);
```
what happens if this is null?

gianm replied:
what happens if YOU are null

gianm replied:
anyway, it's not gonna be null. lookup returns an empty list if it doesn't find anything.

fjy commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1192 @@
+  public <T> QueryRunner<T> getQueryRunnerForSegments(final Query<T> query, final Iterable<SegmentDescriptor> specs)
+  {
+    if (conglomerate == null) {
+      throw new IllegalStateException("Don't query me, bro.");
```
better error msg plz

gianm replied:
no user should ever see this, it means there's a bug.

fjy commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1192 @@
+                                              public QueryRunner<T> apply(final FireHydrant hydrant)
+                                              {
+                                                if (skipIncrementalSegment && !hydrant.hasSwapped()) {
+                                                  return new NoopQueryRunner<>();
```
what is the end experience for hte user when this is returned? Will they understand what is happening?

gianm replied:
only expert users should be setting `skipIncrementalSegment`.

fjy commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1192 @@
+  {
+    // Bail out if this sink is null or otherwise not what we expect.
+    if (sinks.get(identifier) != sink) {
+      log.warn("Sink for segment[%s] no longer valid, bailing out of mergeAndPush.", identifier);
```
what causes this to happen?

gianm replied:
perhaps a sink was dropped and then a different one with the same identifier was re-added.

fjy commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1192 @@
+
+          if (hydrantNumber >= committedHydrants) {
+            log.info("Removing uncommitted segment at [%s]", hydrantDir);
+            FileUtils.deleteDirectory(hydrantDir);
```
what happens if this fails?

fjy replied:
why not call removeDirectory here?

gianm replied:
don't really want the alert, or the log message, or really anything it does other than `FileUtils.deleteDirectory`.

fjy commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1192 @@
+
+        if (committedHydrants <= 0) {
+          log.info("Removing uncommitted sink at [%s]", sinkDir);
+          FileUtils.deleteDirectory(sinkDir);
```
what happens if this fails?

gianm replied:
then the busted directory just stays there on disk, neither being loaded nor deleted.

himanshug replied:
we can log error if deletion fail and continue.

gianm replied:
I think we can't, because we might need to re-use this directory later, and it'd be bad if it already has files in it.

I guess we could skip delete failures here, and then try to delete again later on if we are asked to re-use the directory, and then if *that* fails we can defer failure to that point. but it seems better to fail early, in the unlikely event that we cannot write to our own persist dir.

dclim commented:
If I create a spec that has a starting offset of 1 and then start sending messages, the first message (which has an offset of 0) generates this:
```
2016-02-08T18:00:42,522 INFO [task-runner-0-priority-0] io.druid.indexing.kafka.KafkaIndexTask - Seeking partition[0] to offset[1].
2016-02-08T18:00:42,703 INFO [task-runner-0-priority-0] org.apache.kafka.clients.consumer.internals.Fetcher - Fetch offset null is out of range, resetting offset
2016-02-08T18:02:56,048 INFO [task-runner-0-priority-0] io.druid.segment.realtime.appenderator.AppenderatorImpl - Shutting down...
2016-02-08T18:02:56,051 ERROR [task-runner-0-priority-0] io.druid.indexing.overlord.ThreadPoolTaskRunner - Exception while running task[KafkaIndexTask{id=index_kafka_wikipedia_ffhamgkg, type=index_kafka, dataSource=wikipedia}]
com.metamx.common.ISE: WTF?! Got offset[0] after offset[1] in partition[0].
	at io.druid.indexing.kafka.KafkaIndexTask.run(KafkaIndexTask.java:304) ~[?:?]
	at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:338) [classes/:?]
	at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:318) [classes/:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_60]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_60]
	Suppressed: java.lang.NullPointerException
		at io.druid.segment.realtime.plumber.CoordinatorBasedSegmentHandoffNotifier.close(CoordinatorBasedSegmentHandoffNotifier.java:164) ~[classes/:?]
		at io.druid.segment.realtime.appenderator.FiniteAppenderatorDriver.close(FiniteAppenderatorDriver.java:262) ~[classes/:?]
		at io.druid.indexing.kafka.KafkaIndexTask.run(KafkaIndexTask.java:405) ~[?:?]
		at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:338) [classes/:?]
		at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:318) [classes/:?]
		at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_60]
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_60]
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_60]
		at java.lang.Thread.run(Thread.java:745) [?:1.8.0_60]
2016-02-08T18:02:56,059 INFO [task-runner-0-priority-0] io.druid.indexing.worker.executor.ExecutorLifecycle - Task completed with status: {
  "id" : "index_kafka_wikipedia_ffhamgkg",
  "status" : "FAILED",
  "duration" : 133588
}
```

dclim commented:
If I don't specify "type":"kafka" under ioConfig and tuningConfig, I get the following:
```
{"error":"Unexpected token (END_OBJECT), expected FIELD_NAME: missing property 'type' that is to contain type id  (for class io.druid.indexing.kafka.KafkaIOConfig)\n at [Source: HttpInputOverHTTP@159193cd; line: 1, column: 1959]"}

{"error":"Unexpected token (END_OBJECT), expected FIELD_NAME: missing property 'type' that is to contain type id  (for class io.druid.indexing.kafka.KafkaTuningConfig)\n at [Source: HttpInputOverHTTP@159193cd; line: 1, column: 1559]"}
```

dclim commented:
I'm seeing that even after the segment is built, pushed, and loaded by the historical, the kafka task never completes as the handoff callback is never triggered:
```
2016-02-08T18:47:52,637 INFO [appenderator_merge_0] io.druid.segment.realtime.appenderator.AppenderatorImpl - Pushed merged index for segment[wikipedia6_2013-08-31T00:00:00.000Z_2013-09-01T00:00:00.000Z_2016-02-08T18:46:35.167Z], descriptor is: DataSegment{size=9221, shardSpec=NumberedShardSpec{partitionNum=0, partitions=0}, metrics=[count, added, deleted, delta], dimensions=[page, language, user, unpatrolled, newPage, robot, anonymous, namespace, continent, country, region, city], version='2016-02-08T18:46:35.167Z', loadSpec={type=local, path=/tmp/druid/localStorage/wikipedia6/2013-08-31T00:00:00.000Z_2013-09-01T00:00:00.000Z/2016-02-08T18:46:35.167Z/0/index.zip}, interval=2013-08-31T00:00:00.000Z/2013-09-01T00:00:00.000Z, dataSource='wikipedia6', binaryVersion='9'}
2016-02-08T18:47:52,638 INFO [task-runner-0-priority-0] io.druid.segment.realtime.appenderator.FiniteAppenderatorDriver - Publishing segments with commitMetadata[FiniteAppenderatorDriverMetadata{activeSegments=[wikipedia6_2013-08-31T00:00:00.000Z_2013-09-01T00:00:00.000Z_2016-02-08T18:46:35.167Z], previousSegmentId='wikipedia6_2013-08-31T00:00:00.000Z_2013-09-01T00:00:00.000Z_2016-02-08T18:46:35.167Z', callerMetadata={nextPartitions=KafkaPartitions{topic='wikipedia6', partitionOffsetMap={0=5}}}}]: [DataSegment{size=9221, shardSpec=NumberedShardSpec{partitionNum=0, partitions=0}, metrics=[count, added, deleted, delta], dimensions=[page, language, user, unpatrolled, newPage, robot, anonymous, namespace, continent, country, region, city], version='2016-02-08T18:46:35.167Z', loadSpec={type=local, path=/tmp/druid/localStorage/wikipedia6/2013-08-31T00:00:00.000Z_2013-09-01T00:00:00.000Z/2016-02-08T18:46:35.167Z/0/index.zip}, interval=2013-08-31T00:00:00.000Z/2013-09-01T00:00:00.000Z, dataSource='wikipedia6', binaryVersion='9'}]
2016-02-08T18:47:52,640 INFO [task-runner-0-priority-0] io.druid.indexing.kafka.KafkaIndexTask - Publishing with isTransaction[true].
2016-02-08T18:47:52,640 INFO [task-runner-0-priority-0] io.druid.indexing.common.actions.RemoteTaskActionClient - Performing action for task[index_kafka_wikipedia6_hfaigbfi]: SegmentInsertAction{segments=[DataSegment{size=9221, shardSpec=NumberedShardSpec{partitionNum=0, partitions=0}, metrics=[count, added, deleted, delta], dimensions=[page, language, user, unpatrolled, newPage, robot, anonymous, namespace, continent, country, region, city], version='2016-02-08T18:46:35.167Z', loadSpec={type=local, path=/tmp/druid/localStorage/wikipedia6/2013-08-31T00:00:00.000Z_2013-09-01T00:00:00.000Z/2016-02-08T18:46:35.167Z/0/index.zip}, interval=2013-08-31T00:00:00.000Z/2013-09-01T00:00:00.000Z, dataSource='wikipedia6', binaryVersion='9'}], startMetadata=KafkaDataSourceMetadata{kafkaPartitions=KafkaPartitions{topic='wikipedia6', partitionOffsetMap={0=0}}}, endMetadata=KafkaDataSourceMetadata{kafkaPartitions=KafkaPartitions{topic='wikipedia6', partitionOffsetMap={0=5}}}}
2016-02-08T18:47:52,642 INFO [task-runner-0-priority-0] io.druid.indexing.common.actions.RemoteTaskActionClient - Submitting action for task[index_kafka_wikipedia6_hfaigbfi] to overlord[http://192.168.0.16:8084/druid/indexer/v1/action]: SegmentInsertAction{segments=[DataSegment{size=9221, shardSpec=NumberedShardSpec{partitionNum=0, partitions=0}, metrics=[count, added, deleted, delta], dimensions=[page, language, user, unpatrolled, newPage, robot, anonymous, namespace, continent, country, region, city], version='2016-02-08T18:46:35.167Z', loadSpec={type=local, path=/tmp/druid/localStorage/wikipedia6/2013-08-31T00:00:00.000Z_2013-09-01T00:00:00.000Z/2016-02-08T18:46:35.167Z/0/index.zip}, interval=2013-08-31T00:00:00.000Z/2013-09-01T00:00:00.000Z, dataSource='wikipedia6', binaryVersion='9'}], startMetadata=KafkaDataSourceMetadata{kafkaPartitions=KafkaPartitions{topic='wikipedia6', partitionOffsetMap={0=0}}}, endMetadata=KafkaDataSourceMetadata{kafkaPartitions=KafkaPartitions{topic='wikipedia6', partitionOffsetMap={0=5}}}}
2016-02-08T18:47:52,676 INFO [task-runner-0-priority-0] io.druid.segment.realtime.appenderator.FiniteAppenderatorDriver - Published segments, awaiting handoff.
2016-02-08T18:47:52,677 INFO [task-runner-0-priority-0] io.druid.segment.realtime.plumber.CoordinatorBasedSegmentHandoffNotifier - Adding SegmentHandoffCallback for dataSource[wikipedia6] Segment[SegmentDescriptor{interval=2013-08-31T00:00:00.000Z/2013-09-01T00:00:00.000Z, version='2016-02-08T18:46:35.167Z', partitionNumber=0}]
2016-02-08T18:47:52,677 INFO [task-runner-0-priority-0] io.druid.segment.realtime.appenderator.FiniteAppenderatorDriver - Awaiting handoff...
```
Am I missing a config or something?

gianm commented:
@dclim for the handoff, you might be missing `druid.selectors.coordinator.serviceName`, that has to match your coordinator `druid.service` for handoff to be detected.

gianm commented:
@dclim the whacky errors on not specifying `"type" : "kafka"` are expected and are a jackson thing.

gianm commented:
@dclim I think that WTF is something that I fixed at some point, could you double check that you're using the most recent version of this branch?

dclim commented:
We should prevent the transaction failure exception from being written to the logs when it tries to generate a segment ID that has already been issued, since this seems to happen relatively frequently in normal operation and will succeed on the retry. In the logs, you can see it assign segment ID [wikipedia4_2015-08-31T00:00:00.000Z_2015-09-01T00:00:00.000Z_2016-02-18T18:33:03.273Z_2] to sequenceName [index_kafka_wikipedia4_0(2-4)+2(2-4)], then try and fail to issue the same segment ID to sequenceName [index_kafka_wikipedia4_1(2-4)+3(2-4)]. It then retries and succeeds in issuing segment ID [wikipedia4_2015-08-31T00:00:00.000Z_2015-09-01T00:00:00.000Z_2016-02-18T18:33:03.273Z_3] to sequenceName [index_kafka_wikipedia4_1(2-4)+3(2-4)] 
```
2016-02-18T18:34:31,320 INFO [qtp173783788-78] io.druid.metadata.IndexerSQLMetadataStorageCoordinator - Allocated pending segment [wikipedia4_2015-08-31T00:00:00.000Z_2015-09-01T00:00:00.000Z_2016-02-18T18:33:03.273Z_2] for sequence[index_kafka_wikipedia4_0(2-4)+2(2-4)] (previous = []) in DB
2016-02-18T18:34:31,324 WARN [qtp173783788-59] com.metamx.common.RetryUtils - Failed on try 1, retrying in 2,320ms.
org.skife.jdbi.v2.exceptions.CallbackFailedException: org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: com.mysql.jdbc.exceptions.jdbc4.MySQLIntegrityConstraintViolationException: Duplicate entry 'wikipedia4_2015-08-31T00:00:00.000Z_2015-09-01T00:00:00.000Z_201' for key 'PRIMARY' [statement:"INSERT INTO druid_pendingSegments (id, dataSource, created_date, start, "end", sequence_name, sequence_prev_id, sequence_name_prev_id_sha1, payload) VALUES (:id, :dataSource, :created_date, :start, :end, :sequence_name, :sequence_prev_id, :sequence_name_prev_id_sha1, :payload)", located:"INSERT INTO druid_pendingSegments (id, dataSource, created_date, start, "end", sequence_name, sequence_prev_id, sequence_name_prev_id_sha1, payload) VALUES (:id, :dataSource, :created_date, :start, :end, :sequence_name, :sequence_prev_id, :sequence_name_prev_id_sha1, :payload)", rewritten:"INSERT INTO druid_pendingSegments (id, dataSource, created_date, start, "end", sequence_name, sequence_prev_id, sequence_name_prev_id_sha1, payload) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)", arguments:{ positional:{}, named:{sequence_prev_id:'',payload:[123, 34, 100, 97, 116, 97, 83, 111, 117, 114, 99, 101, 34, 58, 34, 119, 105, 107, 105, 112, 101, 100, 105, 97, 52, 34, 44, 34, 105, 110, 116, 101, 114, 118, 97, 108, 34, 58, 34, 50, 48, 49, 53, 45, 48, 56, 45, 51, 49, 84, 48, 48, 58, 48, 48, 58, 48, 48, 46, 48, 48, 48, 90, 47, 50, 48, 49, 53, 45, 48, 57, 45, 48, 49, 84, 48, 48, 58, 48, 48, 58, 48, 48, 46, 48, 48, 48, 90, 34, 44, 34, 118, 101, 114, 115, 105, 111, 110, 34, 58, 34, 50, 48, 49, 54, 45, 48, 50, 45, 49, 56, 84, 49, 56, 58, 51, 51, 58, 48, 51, 46, 50, 55, 51, 90, 34, 44, 34, 115, 104, 97, 114, 100, 83, 112, 101, 99, 34, 58, 123, 34, 116, 121, 112, 101, 34, 58, 34, 110, 117, 109, 98, 101, 114, 101, 100, 34, 44, 34, 112, 97, 114, 116, 105, 116, 105, 111, 110, 78, 117, 109, 34, 58, 50, 44, 34, 112, 97, 114, 116, 105, 116, 105, 111, 110, 115, 34, 58, 48, 125, 125],start:'2015-08-31T00:00:00.000Z',end:'2015-09-01T00:00:00.000Z',id:'wikipedia4_2015-08-31T00:00:00.000Z_2015-09-01T00:00:00.000Z_2016-02-18T18:33:03.273Z_2',created_date:'2016-02-18T18:34:31.321Z',dataSource:'wikipedia4',sequence_name:'index_kafka_wikipedia4_1(2-4)+3(2-4)',sequence_name_prev_id_sha1:'557BD71C9C18F6FC0514295E107DFE35627A6D25'}, finder:[]}]
	at org.skife.jdbi.v2.DBI.withHandle(DBI.java:284) ~[jdbi-2.63.1.jar:2.63.1]
	at org.skife.jdbi.v2.DBI.inTransaction(DBI.java:329) ~[jdbi-2.63.1.jar:2.63.1]
	at io.druid.metadata.SQLMetadataConnector$3.call(SQLMetadataConnector.java:137) ~[classes/:?]
	at com.metamx.common.RetryUtils.retry(RetryUtils.java:38) [java-util-0.27.7.jar:?]
	at io.druid.metadata.SQLMetadataConnector.retryTransaction(SQLMetadataConnector.java:142) [classes/:?]
	at io.druid.metadata.IndexerSQLMetadataStorageCoordinator.allocatePendingSegment(IndexerSQLMetadataStorageCoordinator.java:371) [classes/:?]
	at io.druid.indexing.common.actions.SegmentAllocateAction.perform(SegmentAllocateAction.java:208) [classes/:?]
	at io.druid.indexing.common.actions.SegmentAllocateAction.perform(SegmentAllocateAction.java:57) [classes/:?]
	at io.druid.indexing.common.actions.LocalTaskActionClient.submit(LocalTaskActionClient.java:64) [classes/:?]
	at io.druid.indexing.overlord.http.OverlordResource$3.apply(OverlordResource.java:280) [classes/:?]
	at io.druid.indexing.overlord.http.OverlordResource$3.apply(OverlordResource.java:269) [classes/:?]
	at io.druid.indexing.overlord.http.OverlordResource.asLeaderWith(OverlordResource.java:518) [classes/:?]
	at io.druid.indexing.overlord.http.OverlordResource.doAction(OverlordResource.java:266) [classes/:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_60]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_60]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_60]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_60]
	at com.sun.jersey.spi.container.JavaMethodInvokerFactory$1.invoke(JavaMethodInvokerFactory.java:60) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.model.method.dispatch.AbstractResourceMethodDispatchProvider$ResponseOutInvoker._dispatch(AbstractResourceMethodDispatchProvider.java:205) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.model.method.dispatch.ResourceJavaMethodDispatcher.dispatch(ResourceJavaMethodDispatcher.java:75) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.HttpMethodRule.accept(HttpMethodRule.java:302) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.ResourceClassRule.accept(ResourceClassRule.java:108) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.RootResourceClassesRule.accept(RootResourceClassesRule.java:84) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1542) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1473) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409) [jersey-servlet-1.19.jar:1.19]
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:558) [jersey-servlet-1.19.jar:1.19]
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:733) [jersey-servlet-1.19.jar:1.19]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) [javax.servlet-api-3.1.0.jar:3.1.0]
	at com.google.inject.servlet.ServletDefinition.doServiceImpl(ServletDefinition.java:278) [guice-servlet-4.0-beta.jar:?]
	at com.google.inject.servlet.ServletDefinition.doService(ServletDefinition.java:268) [guice-servlet-4.0-beta.jar:?]
	at com.google.inject.servlet.ServletDefinition.service(ServletDefinition.java:180) [guice-servlet-4.0-beta.jar:?]
	at com.google.inject.servlet.ManagedServletPipeline.service(ManagedServletPipeline.java:93) [guice-servlet-4.0-beta.jar:?]
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:85) [guice-servlet-4.0-beta.jar:?]
	at com.google.inject.servlet.ManagedFilterPipeline.dispatch(ManagedFilterPipeline.java:120) [guice-servlet-4.0-beta.jar:?]
	at com.google.inject.servlet.GuiceFilter$1.call(GuiceFilter.java:132) [guice-servlet-4.0-beta.jar:?]
	at com.google.inject.servlet.GuiceFilter$1.call(GuiceFilter.java:129) [guice-servlet-4.0-beta.jar:?]
	at com.google.inject.servlet.GuiceFilter$Context.call(GuiceFilter.java:206) [guice-servlet-4.0-beta.jar:?]
	at com.google.inject.servlet.GuiceFilter.doFilter(GuiceFilter.java:129) [guice-servlet-4.0-beta.jar:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) [jetty-servlet-9.2.5.v20141112.jar:9.2.5.v20141112]
	at io.druid.server.http.RedirectFilter.doFilter(RedirectFilter.java:71) [classes/:?]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) [jetty-servlet-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.servlets.UserAgentFilter.doFilter(UserAgentFilter.java:83) [jetty-servlets-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.servlets.GzipFilter.doFilter(GzipFilter.java:364) [jetty-servlets-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) [jetty-servlet-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585) [jetty-servlet-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221) [jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1125) [jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515) [jetty-servlet-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) [jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1059) [jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52) [jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97) [jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.Server.handle(Server.java:497) [jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310) [jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:248) [jetty-server-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540) [jetty-io-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:620) [jetty-util-9.2.5.v20141112.jar:9.2.5.v20141112]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:540) [jetty-util-9.2.5.v20141112.jar:9.2.5.v20141112]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_60]
Caused by: org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: com.mysql.jdbc.exceptions.jdbc4.MySQLIntegrityConstraintViolationException: Duplicate entry 'wikipedia4_2015-08-31T00:00:00.000Z_2015-09-01T00:00:00.000Z_201' for key 'PRIMARY' [statement:"INSERT INTO druid_pendingSegments (id, dataSource, created_date, start, "end", sequence_name, sequence_prev_id, sequence_name_prev_id_sha1, payload) VALUES (:id, :dataSource, :created_date, :start, :end, :sequence_name, :sequence_prev_id, :sequence_name_prev_id_sha1, :payload)", located:"INSERT INTO druid_pendingSegments (id, dataSource, created_date, start, "end", sequence_name, sequence_prev_id, sequence_name_prev_id_sha1, payload) VALUES (:id, :dataSource, :created_date, :start, :end, :sequence_name, :sequence_prev_id, :sequence_name_prev_id_sha1, :payload)", rewritten:"INSERT INTO druid_pendingSegments (id, dataSource, created_date, start, "end", sequence_name, sequence_prev_id, sequence_name_prev_id_sha1, payload) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)", arguments:{ positional:{}, named:{sequence_prev_id:'',payload:[123, 34, 100, 97, 116, 97, 83, 111, 117, 114, 99, 101, 34, 58, 34, 119, 105, 107, 105, 112, 101, 100, 105, 97, 52, 34, 44, 34, 105, 110, 116, 101, 114, 118, 97, 108, 34, 58, 34, 50, 48, 49, 53, 45, 48, 56, 45, 51, 49, 84, 48, 48, 58, 48, 48, 58, 48, 48, 46, 48, 48, 48, 90, 47, 50, 48, 49, 53, 45, 48, 57, 45, 48, 49, 84, 48, 48, 58, 48, 48, 58, 48, 48, 46, 48, 48, 48, 90, 34, 44, 34, 118, 101, 114, 115, 105, 111, 110, 34, 58, 34, 50, 48, 49, 54, 45, 48, 50, 45, 49, 56, 84, 49, 56, 58, 51, 51, 58, 48, 51, 46, 50, 55, 51, 90, 34, 44, 34, 115, 104, 97, 114, 100, 83, 112, 101, 99, 34, 58, 123, 34, 116, 121, 112, 101, 34, 58, 34, 110, 117, 109, 98, 101, 114, 101, 100, 34, 44, 34, 112, 97, 114, 116, 105, 116, 105, 111, 110, 78, 117, 109, 34, 58, 50, 44, 34, 112, 97, 114, 116, 105, 116, 105, 111, 110, 115, 34, 58, 48, 125, 125],start:'2015-08-31T00:00:00.000Z',end:'2015-09-01T00:00:00.000Z',id:'wikipedia4_2015-08-31T00:00:00.000Z_2015-09-01T00:00:00.000Z_2016-02-18T18:33:03.273Z_2',created_date:'2016-02-18T18:34:31.321Z',dataSource:'wikipedia4',sequence_name:'index_kafka_wikipedia4_1(2-4)+3(2-4)',sequence_name_prev_id_sha1:'557BD71C9C18F6FC0514295E107DFE35627A6D25'}, finder:[]}]
	at org.skife.jdbi.v2.SQLStatement.internalExecute(SQLStatement.java:1334) ~[jdbi-2.63.1.jar:2.63.1]
	at org.skife.jdbi.v2.Update.execute(Update.java:56) ~[jdbi-2.63.1.jar:2.63.1]
	at io.druid.metadata.IndexerSQLMetadataStorageCoordinator$3.inTransaction(IndexerSQLMetadataStorageCoordinator.java:552) ~[classes/:?]
	at io.druid.metadata.IndexerSQLMetadataStorageCoordinator$3.inTransaction(IndexerSQLMetadataStorageCoordinator.java:373) ~[classes/:?]
	at org.skife.jdbi.v2.tweak.transactions.LocalTransactionHandler.inTransaction(LocalTransactionHandler.java:184) ~[jdbi-2.63.1.jar:2.63.1]
	at org.skife.jdbi.v2.BasicHandle.inTransaction(BasicHandle.java:327) ~[jdbi-2.63.1.jar:2.63.1]
	at org.skife.jdbi.v2.DBI$5.withHandle(DBI.java:333) ~[jdbi-2.63.1.jar:2.63.1]
	at org.skife.jdbi.v2.DBI.withHandle(DBI.java:281) ~[jdbi-2.63.1.jar:2.63.1]
	... 64 more
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLIntegrityConstraintViolationException: Duplicate entry 'wikipedia4_2015-08-31T00:00:00.000Z_2015-09-01T00:00:00.000Z_201' for key 'PRIMARY'
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_60]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_60]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_60]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422) ~[?:1.8.0_60]
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:377) ~[?:?]
	at com.mysql.jdbc.Util.getInstance(Util.java:360) ~[?:?]
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:971) ~[?:?]
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3887) ~[?:?]
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3823) ~[?:?]
	at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2435) ~[?:?]
	at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2582) ~[?:?]
	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2530) ~[?:?]
	at com.mysql.jdbc.PreparedStatement.executeInternal(PreparedStatement.java:1907) ~[?:?]
	at com.mysql.jdbc.PreparedStatement.execute(PreparedStatement.java:1199) ~[?:?]
	at org.apache.commons.dbcp2.DelegatingPreparedStatement.execute(DelegatingPreparedStatement.java:198) ~[commons-dbcp2-2.0.1.jar:2.0.1]
	at org.apache.commons.dbcp2.DelegatingPreparedStatement.execute(DelegatingPreparedStatement.java:198) ~[commons-dbcp2-2.0.1.jar:2.0.1]
	at org.skife.jdbi.v2.SQLStatement.internalExecute(SQLStatement.java:1328) ~[jdbi-2.63.1.jar:2.63.1]
	at org.skife.jdbi.v2.Update.execute(Update.java:56) ~[jdbi-2.63.1.jar:2.63.1]
	at io.druid.metadata.IndexerSQLMetadataStorageCoordinator$3.inTransaction(IndexerSQLMetadataStorageCoordinator.java:552) ~[classes/:?]
	at io.druid.metadata.IndexerSQLMetadataStorageCoordinator$3.inTransaction(IndexerSQLMetadataStorageCoordinator.java:373) ~[classes/:?]
	at org.skife.jdbi.v2.tweak.transactions.LocalTransactionHandler.inTransaction(LocalTransactionHandler.java:184) ~[jdbi-2.63.1.jar:2.63.1]
	at org.skife.jdbi.v2.BasicHandle.inTransaction(BasicHandle.java:327) ~[jdbi-2.63.1.jar:2.63.1]
	at org.skife.jdbi.v2.DBI$5.withHandle(DBI.java:333) ~[jdbi-2.63.1.jar:2.63.1]
	at org.skife.jdbi.v2.DBI.withHandle(DBI.java:281) ~[jdbi-2.63.1.jar:2.63.1]
	... 64 more
2016-02-18T18:34:31,678 INFO [qtp173783788-53] io.druid.indexing.common.actions.LocalTaskActionClient - Performing action for task[index_kafka_wikipedia4_0(2-4)+2(2-4)_kiehognj]: SegmentInsertAction{segments=[DataSegment{size=9221, shardSpec=NumberedShardSpec{partitionNum=2, partitions=0}, metrics=[count, added, deleted, delta], dimensions=[page, language, user, unpatrolled, newPage, robot, anonymous, namespace, continent, country, region, city], version='2016-02-18T18:33:03.273Z', loadSpec={type=local, path=/Users/david/code/druid-tmp/druid/localStorage/wikipedia4/2015-08-31T00:00:00.000Z_2015-09-01T00:00:00.000Z/2016-02-18T18:33:03.273Z/2/index.zip}, interval=2015-08-31T00:00:00.000Z/2015-09-01T00:00:00.000Z, dataSource='wikipedia4', binaryVersion='9'}], startMetadata=null, endMetadata=null}
2016-02-18T18:34:31,678 INFO [qtp173783788-53] io.druid.indexing.overlord.MetadataTaskStorage - Logging action for task[index_kafka_wikipedia4_0(2-4)+2(2-4)_kiehognj]: SegmentInsertAction{segments=[DataSegment{size=9221, shardSpec=NumberedShardSpec{partitionNum=2, partitions=0}, metrics=[count, added, deleted, delta], dimensions=[page, language, user, unpatrolled, newPage, robot, anonymous, namespace, continent, country, region, city], version='2016-02-18T18:33:03.273Z', loadSpec={type=local, path=/Users/david/code/druid-tmp/druid/localStorage/wikipedia4/2015-08-31T00:00:00.000Z_2015-09-01T00:00:00.000Z/2016-02-18T18:33:03.273Z/2/index.zip}, interval=2015-08-31T00:00:00.000Z/2015-09-01T00:00:00.000Z, dataSource='wikipedia4', binaryVersion='9'}], startMetadata=null, endMetadata=null}
2016-02-18T18:34:31,688 INFO [qtp173783788-53] io.druid.metadata.IndexerSQLMetadataStorageCoordinator - Published segment [wikipedia4_2015-08-31T00:00:00.000Z_2015-09-01T00:00:00.000Z_2016-02-18T18:33:03.273Z_2] to DB
2016-02-18T18:34:33,669 INFO [qtp173783788-59] io.druid.metadata.IndexerSQLMetadataStorageCoordinator - Allocated pending segment [wikipedia4_2015-08-31T00:00:00.000Z_2015-09-01T00:00:00.000Z_2016-02-18T18:33:03.273Z_3] for sequence[index_kafka_wikipedia4_1(2-4)+3(2-4)] (previous = []) in DB
```

gianm commented:
Updated the top comment with some more details and a couple of discussion items I'd like feedback on.

Also removed the [wip] tag as the TODOs are resolved and things work to the best of my knowledge. We have started running this in our cluster yesterday and are keeping an eye on that for how well it's working (looks ok so far)

himanshug commented on server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java:
```diff
@@ -264,24 +270,90 @@ public void start()
+                if (!success) {
+                  transactionStatus.setRollbackOnly();
+                  txnFailure.set(true);
+                  return null;
```
this should return `new SegmentPublishResult(ImmutableList.of(), false)` instead of null

gianm replied:
this return value is not used, as `transactionStatus.setRollbackOnly()` causes the txn to throw an exception rather than return a value. will make a note of this in a comment.

gianm replied:
actually I'll just make this throw an exception.

himanshug commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1191 @@
+      // Try one more time after swapping, then throw the exception out if it happens again.
+      persistAll(committerSupplier.get());
+      sinkRetVal = sink.add(row);
+    }
```
we need another catch block here to honor "reportParseExceptions"

gianm replied:
`sink.add` won't throw parse exceptions if `reportParseExceptions` is off, as they are suppressed by the incremental index. actually, I think we should probably adjust Plumbers to not include sink.add in the try/catch there, so will take a look at that in a separate PR.

gianm replied:
#2631

himanshug commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -0,0 +1,1191 @@
+      log.makeAlert("Unknown query type, [%s]", query.getClass())
+         .addData("dataSource", query.getDataSource())
+         .emit();
+      return new NoopQueryRunner<>();
```
this should probably be an exception instead of responding with empty result.

gianm replied:
sounds good, will change to ISE

fjy commented:
:+1:

himanshug commented on server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java:
```diff
@@ -0,0 +1,521 @@
+    handoffNotifier.start();
+
+    final FiniteAppenderatorDriverMetadata metadata = objectMapper.convertValue(
+        appenderator.startJob(),
```
appenderator.startJob() might give null in many cases.

gianm replied:
that's ok, convertValue turns null into null, and then there's a null check in this method.

himanshug replied:
ok, dint know that converValue could handle null.

himanshug commented on server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java:
```diff
@@ -0,0 +1,521 @@
+
+    if (segmentsAndMetadata != null) {
+      log.info("Awaiting handoff...");
+      synchronized (handoffMonitor) {
```
not a blocker: but can this honor handoffConditionTimeout ?

gianm replied:
yes, I can port that here, as ideally this supports every feature that RealtimePlumber supports.

gianm replied:
ok, added. the semantics are slightly different than in RealtimePlumber- here it only applies to historicals loading the data, not to push/publish. but I think this is actually more useful anyway, as the historicals are probably the thing we want to be timing out on, rather than push to deep storage.

himanshug commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java:
```diff
@@ -0,0 +1,135 @@
+      @JsonProperty("buildV9Directly") Boolean buildV9Directly,
+      @JsonProperty("reportParseExceptions") Boolean reportParseExceptions
+  )
+  {
```
lots of duplication among various TuningConfig impls, does it make sense to have an abstract BaseTuningConfig or at least a BaseRealtimeTuningConfig because RealtimeTuningConfig and KafkaTuningConfig look very same?

gianm replied:
the AppenderatorConfig is sorta a base interface they both share, although it does not have implementation. do you think it would be better for it to have implementation (and Jackson annotations) too?

gianm replied:
I think it's probably not the worst thing in the world, the code's not *that* long and it makes it clear which options are actually respected by which tasks. imo it would be unfortunate to get into a situation where some options are just ignored sometimes.

but I am kinda on the fence, and could be convinced otherwise.

himanshug replied:
we can keep it this way for now.

himanshug commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java:
```diff
@@ -0,0 +1,584 @@
+  public <T> QueryRunner<T> getQueryRunner(Query<T> query)
+  {
+    if (appenderator == null) {
+      return null;
```
should we return NoopQueryRunner here, returning null here might fail the druid query, whereas its just that there is no data to respond at this point from this task.

gianm replied:
sounds good, changed.

himanshug commented:
@gianm had some comments but overall looks good and from what I see this PR does not change  anything in RealtimeIndexTask  so it should have no impact on existing ingestion mechanism.

regarding the discussion points..
1) I think we can update RealtimePlumber  at some point if/when possible.
2) I'm not sure if we need to update IndexGeneratorReducer with this.

gianm commented:
@himanshug updated with most of your review comments addressed, but I am not entirely sure what you mean on the thread w/ https://github.com/druid-io/druid/pull/2220#discussion_r55778213.

himanshug commented:
:+1: @gianm looks good overall, can you pls cleanup your commit history as needed.

gianm commented:
@himanshug I had it split up as 6 commit on purpose (I think the 6 commits are pretty distinct), do you think I should merge some of them?

himanshug commented:
i did say, "as needed" in there. if its already done then, great. :+1:

gianm commented:
cool, just checking :smile:
